src/cod_parser.cpp

```diff
@@ -0,0 +1,912 @@
+    {
+      if (current_object->variables().variable(index).name() == variable_name)
+      {
+        return index;
+      }
+    }
+  }
+  return -1;
+}
+
+cod_pb::Variable* Cod_Parser::create_or_reuse_variable(const std::string& name)
+{
+  if (current_object)
+  {
+    auto optional_var = get_variable(current_object, name);
+    if (optional_var)
+    {
+      return optional_var.value();
+    }
+  }
+  return current_object->mutable_variables()->add_variable();
+}
+
+cod_pb::Variable Cod_Parser::get_variable(const std::string& key)
+{
+  for (int i = 0; i < constants.variable_size(); i++)
+  {
+    if (constants.variable(i).name() == key)
+    {
+      return constants.variable(i);
+    }
+  }
+  cod_pb::Variable ret;
+  return ret;
+}
+
+std::experimental::optional<cod_pb::Variable*> Cod_Parser::get_variable(cod_pb::Object* obj, const std::string& name)
+{
+  for (int i = 0; i < obj->variables().variable_size(); i++)
+  {
+    if (obj->variables().variable(i).name() == name)
+    {
+      return obj->mutable_variables()->mutable_variable(i);
+    }
+  }
+  return {};
+}
+
+int Cod_Parser::calculate_operation(int old_value, const std::string& operation, int op)
+{
+  int current_value = old_value;
+  if (operation == "+")
+  {
+    current_value += op;
+  }
+  else if (operation == "-")
+  {
+    current_value -= op;
+  }
+  else
+  {
+    current_value = op;
+  }
+  return current_value;
+}
+
+
+// Object stack related functions
+bool Cod_Parser::top_is_number_object()
+{
+  if (!object_stack.empty())
+  {
+    return object_stack.top().number_object;
+  }
+  return false;
+}
+
+void Cod_Parser::add_object_to_stack(cod_pb::Object* o, bool number_object, int spaces)
+{
+  ObjectType obj;
+  obj.object = o;
+  obj.number_object = number_object;
+  obj.spaces = spaces;
+  object_stack.push(obj);
+}
+
+bool Cod_Parser::object_finished()
+{
+  if (object_stack.size() > 0)
+  {
+    object_stack.pop();
+  }
+}
+
+// String handling functions
+std::vector<std::string> Cod_Parser::regex_match(std::string regex, std::string str)
+{
+  std::vector<std::string> ret;
+  boost::regex expr{regex};
+  boost::smatch what;
+  if (boost::regex_match(str, what, expr))
+  {
+    for (int i = 0; i < what.size(); i++)
+    {
+      ret.push_back(what[i]);
+    }
+  }
+  return ret;
+}
+
+std::vector<std::string> Cod_Parser::regex_search(std::string regex, std::string str)
+{
+  std::vector<std::string> ret;
+  boost::regex expr{regex};
+  boost::smatch what;
+  if (boost::regex_search(str, what, expr))
+  {
+    for (int i = 0; i < what.size(); i++)
+    {
+      ret.push_back(what[i]);
+    }
+  }
+  return ret;
+}
+
+
+std::string Cod_Parser::tabs_to_spaces(const std::string& str)
+{
+  std::string newtext = "  ";
+  boost::regex re("\t");
+
+  std::string result = boost::regex_replace(str, re, newtext);
+  return result;
+}
+
+int Cod_Parser::count_front_spaces(const std::string& str)
```

Comment:
Warum spielt es eine Rolle, wie weit es eingerückt ist?

Reply:
Das wird verwendet um die, ich nenne sie, Nummernobjekte abzuschliessen, wenn es das letzte in einem Objekt ist. 
Beispiel:
```
Objekt: OBJEKTS
   Objekt: OBJ
      Nummer: 0
      var: 1
      Nummer: 2
      var: 2         
   EndObj     <- Obekt Nummer 2 wird geschlossen, da ein EndObj + die Spaces < als zuvor.
EndObj
```